lab2/lab2.md

```diff
@@ -1,30 +1,624 @@
+|--------|------------|------|-----|---------|-----------|--------|---------|--------|
+|imm[12] | imm[10:5]  | rs2  | rs1 | funct3  | imm[4:1]  | imm[11]|  opcode | B-type |
+|--------|------------|------|-----|---------|-----------|--------|---------|--------|
+|     imm[12|10:5]]   | rs2  | rs1 |   000   |     imm[4:1|11]    | 1100011 |  BEQ   |
+|     imm[12|10:5]]   | rs2  | rs1 |   001   |     imm[4:1|11]    | 1100011 |  BNE   |
+|     imm[12|10:5]]   | rs2  | rs1 |   100   |     imm[4:1|11]    | 1100011 |  BLT   |
+|     imm[12|10:5]]   | rs2  | rs1 |   101   |     imm[4:1|11]    | 1100011 |  BGE   |
+|     imm[12|10:5]]   | rs2  | rs1 |   110   |     imm[4:1|11]    | 1100011 |  BLTU  |
+|     imm[12|10:5]]   | rs2  | rs1 |   111   |     imm[4:1|11]    | 1100011 |  BGEU  |
+
+
+This table is from the RISC-V User-level ISA Specification v2.2, page 104.
+You can find the same information in Chapter 2 of the Specification, Chapter 2 of the RISC-V reader, or in the front of the Computer Organization and Design book.
+
+
+Given these inputs, you must generate the correct output on the taken wire.
+The template code from `src/main/scala/components/branch-control.scala` is shown below.
+You will fill in where it says *Your code goes here*.
+
+```
+// Control logic for whether branches are taken or not
+
+package dinocpu
+
+import chisel3._
+import chisel3.util._
+
+/**
+ * Controls whether or not branches are taken.
+ *
+ * Input:  branch, true if we are looking at a branch
+ * Input:  funct3, the middle three bits of the instruction (12-14). Specifies the
+ *         type of branch
+ * Input:  inputx, first value (e.g., reg1)
+ * Input:  inputy, second value (e.g., reg2)
+ * Output: taken, true if the branch is taken.
+ */
+class BranchControl extends Module {
+  val io = IO(new Bundle {
+    val branch = Input(Bool())
+    val funct3 = Input(UInt(3.W))
+    val inputx = Input(UInt(32.W))
+    val inputy = Input(UInt(32.W))
+
+    val taken  = Output(Bool())
+  })
+
+// Your code goes here
+
+  io.taken := false.B
+}
+
+}
+```
+
+**HINT:** Use Chisel's `switch` / `is`,  `when` / `elsewhen` / `otherwise`, or `MuxCase` syntax.
+See [the Chisel getting started guide](../chisel-notes/getting-started.md) for examples.
+You may also find the [Chisel cheat sheet](https://chisel.eecs.berkeley.edu/2.2.0/chisel-cheatsheet.pdf) helpful.
+
+### Testing your branch-control unit
+
+We have implemented some tests for your barnch-control unit. The tests along with the other lab-2 tests are in `src/test/scala/labs/Lab2Test.scala`.
+
+In this part of the assignment, you only need to run the branch-control unit tests.
+To run just these tests, you can use the sbt command `testOnly`, as demonstrated below.
+
+
+```
+dinocpu:sbt> testOnly Lab2 / dinocpu.BranchControlTesterLab2
+```
+
+Feel free to add your own tests in `src/tests/scala`, modify the current tests, and add `print` statements in the tests.
+
+## Draw a diagram for implementing the branch-control unit
```

Comment:
You mean for implementing branch instructions???

I think we want them to just turn in one diagram which as everything. However, it's going to be pretty messy.

The key things I want to see in the diagram is that they have the correct muxes in the correct places. We should brainstorm this a bit. @jlperona any ideas?